### PR TITLE
simplefs: Cancelling v1

### PR DESCRIPTION
**Pull request by taruti**
As discussed on Slack.

### CODE REVIEW

File: simplefs/simplefs.go [hunk 1]
```diff
@@ -517,6 +542,20 @@ func (k *SimpleFS) SimpleFSClose(ctx context.Context, opid keybase1.OpID) (err e
...
+	k.lock.Lock()
+	defer k.lock.Unlock()
+	w, ok := k.inProgress[opid]
+	if !ok {
+		return
+	}
+	delete(k.inProgress, opid)
+	delete(k.handles, opid)
+	w.cancel()
+}
```
Comment (oconnor663) [inline simplefs/simplefs.go]: Will this work for `keybase fs write`, for example? Or would we still need to call `Close` there?

Comment (taruti) [inline simplefs/simplefs.go]: k.handles is what Close clears, so after a Cancel no Close or Wait needed.

Comment (oconnor663) [inline simplefs/simplefs.go]: Cool. Is there nothing behind the `Node` in the handle that needs to get closed / cleaned up? Just the `Node` itself?

File: simplefs/simplefs.go [hunk 2]
```diff
@@ -219,10 +220,28 @@ func (k *SimpleFS) doCopy(ctx context.Context, srcPath, destPath keybase1.Path)
...
+		if err := ctx.Err(); err != nil {
+			return err
+		}
+		_, err := io.CopyN(dst, src, 64*1024)
+		if err == io.EOF {
+			return nil
+		}
+		if err != nil {
+			return err
+		}
```
Comment (zanderz) [inline simplefs/simplefs.go]: `err == nil` always means there is more to copy?

Comment (oconnor663) [inline simplefs/simplefs.go]: I think the implementation of `Reader` is allowed to choose whether to return `EOF` with the last non-zero-length read, or to return it on the first zero-length read. But either way, I think it's correct for callers to loop until `EOF` like this.

File: simplefs/simplefs.go [hunk 3]
```diff
@@ -517,6 +542,20 @@ func (k *SimpleFS) SimpleFSClose(ctx context.Context, opid keybase1.OpID) (err e
 	return err
 }
 
+// SimpleFSCancel starts to cancel op with the given opid.
+// Also remove any pending references of opid everywhere.
+func (k *SimpleFS) SimpleFSCancel(_ context.Context, opid keybase1.OpID) {
```
Comment (oconnor663) [inline simplefs/simplefs.go]: The generated protocol is going to expect a method that returns `error` I think.

**Comment by oconnor663** [comment]
LGTM with the error change above. It looks like this is working on my example branches, https://github.com/keybase/client/tree/jack/CORE-4684/cancel_scratch and https://github.com/keybase/kbfs/tree/jack/CORE-4684/cancel_scratch.

File: simplefs/simplefs.go (hunk 1 above)
Comment (taruti) [inline simplefs/simplefs.go]: Node does not need closing.

File: simplefs/simplefs.go (hunk 2 above)
Comment (taruti) [inline simplefs/simplefs.go]: This uses CopyN for which ```On return, written == n if and only if err == nil.```.

However on reflection using CopyBuffer is better for the given types.

File: simplefs/simplefs.go (hunk 3 above)
Comment (taruti) [inline simplefs/simplefs.go]: Will do.

**Pull request merged**